### PR TITLE
fix(expr): fix `coalesce` display problem 

### PR DESCRIPTION
**Pull request by likzn**
### Motivation
The old logic will select first `ColumnType` in `Vec` as output, but in below case the first `ColumnType`'s `scalar type` is `Char` and its length is 1 so it can not display correctly. I fix it with adding `filter` to avoid this problem(We only select the first `Not Null Expr`). Relate issue: #11166
```
materialize=> select coalesce(null::char(1), 'abc');
 coalesce 
----------
 a
(1 row)
```

And I fix a doc link problem by the way when i searched the way to test :)

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -5592,8 +5591,11 @@ impl VariadicFunc {
                     "coalesce/greatest/least inputs did not have uniform type: {:?}",
                     input_types
                 );
-                let nullable = input_types.iter().all(|ty| ty.nullable);
-                input_types.into_first().nullable(nullable)
+                input_types
+                    .iter()
+                    .find_or_first(|ty| !ty.nullable)
+                    .unwrap()
```
Comment (likzn) [inline src/expr/src/scalar/func.rs:5597]: Line in 5586 had a `assert` with `input_types.len()`. So this `unwrap` is safe.

**Comment by likzn** [comment]
@benesch @philip-stoev Hi, PTAL~

**Comment by benesch** [comment]
You were right about where the bug was, it turns out! I've pushed up a fix to have coalesce and friends union together all types—and also adjusted the definition of type union to discard modifiers. That should do the trick.

**Comment by benesch** [comment]
Thanks again for the PR!

**Comment by benesch** [comment]
@philip-stoev are you comfortable with me adjusting a few of the tests in char-varchar-distinct.td? The problem is that because we don't support [data types with complex comparison semantics](https://github.com/MaterializeInc/materialize/issues/10566) we don't have a way to match PostgreSQL's behavior exactly. The best we can do is to drop the char type modifier after a `DISTINCT` operation, allowing each value to be exactly as wide as its contents, with no enforced trailing space padding. I think it is worth accepting the small regression in char-varchar-distinct.td in exchange for consistent behavior across all homogenizing functions. In particular, it seems more important that `select coalesce(null::char(1), 'abc');` not truncate to `abc` than it does for `DISTINCT` on `char` columns to preserve trailing spaces.

**Comment by philip-stoev** [comment]
@benesch I am OK with a tradeoff if you think it is better. Should we document it somehow in the documentation?

The thing with CHAR is that A) people continue to have such columns in their tables, B) those columns continue to have random trailing spaces in them, as if the data is still coming from a mainframe with punched cards and C) those trailing spaces will raise their ugly head one way or another. So expect some more of this once we obtain a critical mass of customers.

**Comment by benesch** [comment]
> @benesch I am OK with a tradeoff if you think it is better.

Great. The changes were quite minor. I noted them all in the testdrive script.

> Should we document it somehow in the documentation?

Ideally yes, but we don't document our support for the `char` type at all right now!

> The thing with CHAR is that A) people continue to have such columns in their tables, B) those columns continue to have random trailing spaces in them, as if the data is still coming from a mainframe with punched cards and C) those trailing spaces will raise their ugly head one way or another. So expect some more of this once we obtain a critical mass of customers.

Noted. I do hope we can address this properly some day with #10566. It is just a monumental amount of work. 🤷🏽‍♂️

**Pull request merged**